opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -372,37 +374,57 @@ def set_attribute(self, key: str, value: types.AttributeValue) -> None:
             logger.warning("invalid key (empty or null)")
             return
 
-        if isinstance(value, Sequence):
-            error_message = self._check_attribute_value_sequence(value)
-            if error_message is not None:
-                logger.warning("%s in attribute value sequence", error_message)
-                return
-            # Freeze mutable sequences defensively
-            if isinstance(value, MutableSequence):
-                value = tuple(value)
-        elif not isinstance(value, (bool, str, int, float)):
-            logger.warning("invalid type for attribute value")
+        error_message = self._check_attribute_value(value)
+        if error_message is not None:
+            logger.warning(error_message)
             return
+        # Freeze mutable sequences defensively
+        if isinstance(value, MutableSequence):
+            value = tuple(value)
 
         self.attributes[key] = value
 
     @staticmethod
-    def _check_attribute_value_sequence(sequence: Sequence) -> Optional[str]:
+    def _check_attribute_value(value: types.AttributeValue) -> Optional[str]:
         """
         Checks if sequence items are valid and are of the same type
         """
-        if len(sequence) == 0:
-            return None
+        valid_types = (bool, str, int, float)
+        if isinstance(value, Sequence):
+            if len(value) == 0:
+                return None
 
-        first_element_type = type(sequence[0])
+            first_element_type = type(value[0])
 
-        if first_element_type not in (bool, str, int, float):
-            return "invalid type"
+            if first_element_type not in valid_types:
+                return "Invalid type {} in attribute value sequence. Expected one of {}".format(
+                    first_element_type.__name__,
+                    [valid_type.__name__ for valid_type in valid_types],
+                )
+
+            for element in value:
+                if not isinstance(element, first_element_type):
+                    return "Mixed types {} and {} in attribute value sequence".format(
+                        first_element_type.__name__, type(element).__name__
+                    )
+            return None
+        elif not isinstance(value, valid_types):
+            return "Invalid type {} for attribute value. Expected one of {}".format(
+                type(value).__name__,
+                [valid_type.__name__ for valid_type in valid_types],
+            )
 
-        for element in sequence:
-            if not isinstance(element, first_element_type):
-                return "different type"
-        return None
+    def _filter_attribute_values(self, attributes: types.Attributes):
+        if attributes:
+            for attr_key, attr_value in list(attributes.items()):
```

Comment:
How come you need the `list()`?

Reply:
The dict (can be) change in the loop, that isn't allowed if you use a generator / lazy loop to go through the dict.